handout/di/angular2/avoiding_collisions_opaque_token.md

```diff
@@ -0,0 +1,98 @@
+
+Consider this example where the main app is a consumer of two modules: one that provides an email service and another that provides a logging service.
+
+_app/email/email.service.ts_
+```typescript
+export const apiConfig = 'api-config';
+
+@Injectable()
+export class EmailService {
+  constructor(@Inject(apiConfig) public apiConfig) { }
+}
+```
+
+_app/email/email.module.ts_
+```typescript
+@NgModule({
+  providers: [ EmailService ],
+})
+export class EmailModule { }
+```
+
+The email service has an associated api that is trying to inject configuration settings identified by the string `'api-config'`.
+
+_app/logger/logger.service.ts_
+```typescript
+export const apiConfig = 'api-config';
+
+@Injectable()
+export class LoggerService {
+  constructor(@Inject(apiConfig) public apiConfig) { }
+}
+```
+
+_app/logger/logger.module.ts_
+```typescript
+@NgModule({
+  providers: [ LoggerService ],
+})
+export class LoggerModule { }
+```
+
+`LoggerModule` has an api as well, and it's also using the string `'api-config'` to inject its configuration settings. When it comes time for the main app to specify those settings, Angular overwrites the first `emailApiConfig` value with the `loggerApiConfig` value, since it was provided last.
+
+_app/app.module.ts_
+```typescript
+import { apiConfig as emailApiConfig } from './email/index';
+import { apiConfig as loggerApiConfig } from './logger/index';
+
+@NgModule({
+  ...
+  providers: [
+    { provide: emailApiConfig, useValue: { apiKey: 'email-config', context: 'registration' } },
+    { provide: loggerApiConfig, useValue: { apiKey: 'logger-config' } },
+  ],
+  ...
+})
+export class AppModule { }
+```
+[View Example](https://plnkr.co/edit/QrvjsucT6lF6dnFUb2ag?p=preview)
+
+In this case, module implementation details are leaking out to the parent module. Not only that, those details were obfuscated through the module exports and this can lead to problematic debugging. This is where Angular's `OpaqueToken` comes into play.
+
+## OpaqueToken
+
+`OpaqueToken`s are unique and immutable values which allow developers to avoid collisions of dependency injection token ids.
+
+```typescript
+import { OpaqueToken } from '@angular/core';
+
+const name = 'token';
+const token1 = new OpaqueToken(name);
+const token2 = new OpaqueToken(name);
+
+console.log(token1 === token2); // false
+```
+
+Here, regardless of whether or not the same value is passed to the constructor of the token, it will not result in identical symbols.
+
+Replacing the old strings with `OpaqueToken`s allows Angular to resolve entities as both modules intended.
+
+_app/email/email.module.ts_
+```typescript
+export const apiConfig = new OpaqueToken('api-config');
+
+@Injectable()
+export class EmailService {
+  constructor(@Inject(apiConfig) public apiConfig: EmailConfig) { }
+}
+
+@NgModule({
+  providers: [ EmailService ],
+})
+export class EmailModule { }
+```
+[View Example](https://plnkr.co/edit/SHfTH9R6JVDwJKnzRFSH?p=preview)
```

Comment:
I think we should only show the refactoring of the services where the tokens are defined and add some final conclusions. Proposed changes below:

---
_app/email/email.module.ts_

```ts
import { Injectable, Inject, OpaqueToken } from '@angular/core';
import { EmailConfig } from './email.config';

export const apiConfig = new OpaqueToken('api-config');

@Injectable()
export class EmailService {
  constructor(@Inject(apiConfig) public apiConfig: EmailConfig) { }
}
```

_app/logger/logger.service.ts_

```ts
import { Injectable, Inject, OpaqueToken } from '@angular/core';
import { LoggerConfig } from './logger.config';

export const apiConfig = new OpaqueToken('api-config');

@Injectable()
export class LoggerService {
  constructor(@Inject(apiConfig) public apiConfig: LoggerConfig) { }
}
```

[View Example](https://plnkr.co/edit/SHfTH9R6JVDwJKnzRFSH?p=preview)

The rest of the code stays exactly the same as before but now, the collision is avoided. Every service gets the correct configuration object from the root module. With `OpaqueToken`, Angular is now able to differentiate two tokens that uses the same string.